rs-matter/src/respond.rs

```diff
@@ -0,0 +1,269 @@
+/*
+ *
+ *    Copyright (c) 2020-2022 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+use core::fmt::Display;
+use core::pin::pin;
+
+use embassy_futures::select::{select3, select_slice};
+
+use log::{error, info};
+
+use crate::data_model::core::{DataModel, IMBuffer};
+use crate::data_model::objects::DataModelHandler;
+use crate::data_model::subscriptions::Subscriptions;
+use crate::error::{Error, ErrorCode};
+use crate::interaction_model::busy::BusyInteractionModel;
+use crate::secure_channel::busy::BusySecureChannel;
+use crate::secure_channel::core::SecureChannel;
+use crate::transport::exchange::Exchange;
+use crate::utils::buf::BufferAccess;
+use crate::utils::select::Coalesce;
+use crate::Matter;
+
+/// A trait modeling a generic handler for an exchange.
+pub trait ExchangeHandler {
+    async fn handle(&self, exchange: &mut Exchange<'_>) -> Result<(), Error>;
+
+    fn compose<T>(self, other: T) -> CompositeExchangeHandler<Self, T>
```

Comment:
would `chain` be a more common way of saying you are sequencing several items together? I expect composing f and g to be f(g(input)) where as chaining seems to be run f then g.

Reply:
reading more through the doc, this goes through each item and expects either full handling or a protocol-error.

Maybe then `alternate` is a typical name for picking one of several choices? I have some difficulty naming this. If nothing is obviously better feel free to ignore this comment. It just felt a bit off as compose felt like `self(other(input))`

Reply:
I'll replace the `compose` terminology with `chain`. Chain does not resonate with me that much either, but we have adopted it elsewhere - for the `AsyncHandler`s ... "chaining", so we better use the same terminology everywhere.

Reply:
I've done that now. The terminology used w.r.t. `Handler` is "chain", and the struct is `ChainedHandler`, hence I changed all terminology used w.r.t. `ExchangeHandler` to use the word "chain" as well, and the struct to be named `ChanedExchangeHandler`.

More importantly, rather than failing with an error, I've re-implemented the exchange handlers' chaining in a similar way to how it is done for cluster chaining - when chaining two handlers together, the _user_ is responsible for providing which protocol ID will be handled by the first handler.

A similar naming and approach will be soon utilized for `NetworkSend` and `NetworkReceive`, where we need to chain these so as to be able to handle simultaneously UDP, TCP and BTP which - from the pov of the Transport layer, would all look like a single pair of `NetworkSend` and `NetworkReceive` capable of handling all three protocols (or two of the three or just one - up to the user how she would chain the send/receive impls).